Step3 구현

isisrin opened the pull request:
- AnswerDto 추가
- Answer : Builder 추가 및 @tostring추가
- Question : @JsonIgnore추가
- QuestionDto 이름 바꿈
- QnaServiceTest 내용 보완

isisrin committed:
step3 피드백반영 중... 의심가는 애들 커밋
 - @Builder 추가
 - validate함수 구분
 - isNotOwner 추가

isisrin committed:
step3 피드백반영 중... 의심가는 애들 커밋
 - QnaServiceTest추가

isisrin committed:
step3 피드백반영 구현중... 전체커밋

isisrin committed:
step3 구현
 - AnswerDto 추가
 - Answer : Builder 추가 및 @ToString추가
 - Question : @JsonIgnore추가
 - QuestionDto 이름 바꿈
 - QnaServiceTest 내용 보완

isisrin committed:
step3 구현 - 안올라간 것 추가

jojoldu commented on src/main/java/nextstep/domain/Answer.java:
```diff
@@ -40,25 +44,42 @@ public Answer(Long id, User writer, Question question, String contents) {
 
-    public boolean isOwner(User loginUser) {
-        return writer.equalsNameAndEmail(loginUser);
+    public boolean isNotOwner(User loginUser) {
```
피드백 반영 💯

jojoldu commented on src/main/java/nextstep/domain/RequestAnswerDto.java:
```diff
@@ -0,0 +1,19 @@
+    private Question question;
+    
+    public Answer toEntity() {
+        return Answer.builder().writer(writer).question(question).contents(contents).build();
```
가로줄이 너무 긴데요.

```java
Answer.builder()
.writer(writer)
.question(question)
.contents(contents)
.build();
```

로 수정하시면 보기 좋을것 같습니다.

jojoldu commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -15,38 +16,39 @@
 import java.util.Optional;
 
 @Service("qnaService")
+@RequiredArgsConstructor
```
피드백 반영 💯

jojoldu commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -15,38 +16,39 @@
 @Service("qnaService")
+@RequiredArgsConstructor
 public class QnaService {
     private static final Logger log = LoggerFactory.getLogger(QnaService.class);
```
이 로그는 롬복의 @Slf4j 로 해결 가능합니다 :)

jojoldu commented on src/test/java/nextstep/service/QnaServiceTest.java:
```diff
@@ -6,130 +6,101 @@
 
-@RunWith(MockitoJUnitRunner.Silent.class)
+@RunWith(SpringRunner.class)
+@SpringBootTest
```
이렇게 SpringBootTest를 사용할때 주의할점이 하나 있는데요.
**전체 테스트를 한번에 돌릴때 테스트 간에 서로 간섭하지 않도록**해야 합니다.
간섭이라는게 어떤 의미냐면, 
전체 테스트를 돌릴때 SpringBoot는 H2 하나를 실행해 해당 H2만으로 전체 테스트를 수행합니다.

즉, A라는 테스트에서 **Question 테이블에 데이터를 넣었다면** B 테스트 실행시 **Question 테이블에는 A 테스트에서 넣은 데이터가 존재한 채로** 실행됩니다.
이렇게 되면 테스트 검증시에 먼저 실행된 테스트로 데이터가 등록된게 남아 있어 실패할 확률이 높습니다.

해결책은 @After를 이용하여 **테스트가 끝나면 항상 사용된 Repository를 비워주는 것**입니다.
repository.deleteAll() 메소드를 이용하시면 됩니다.

지금같은 경우 Question과 Answer 등등 사용된 테이블들은 항상 비워줄수있도록 @After 메소드를 추가해보세요 :)

jojoldu commented on src/test/java/nextstep/web/api/ApiQnaAcceptanceTest.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    @Test
+    public void 답변_실패() {
```
메소드명을 좀 더 명확하게 지으시면 좋을것 같습니다 :)
**어떤 필드가 없으면 답변 실패** 등의 이름으로 지으시면 **다른 분이 테스트 코드 볼때 한눈에 알것 같습니다**

jojoldu reviewed (changes requested):
@isisrin 혜린님 고생하셨습니다 :)
제가 피드백 드린것 전부 다 반영해주셔서 감사합니다.

이번엔 조금은 부가적인 내용들을 피드백으로 남겼습니다.
참고하시어 수정해주세요 :)

isisrin committed:
step3 피드백 반영
 - 안쓰는 생성자 삭제 Question
 - builder 여러줄로 바꿈
 - Slf4j 추가
 - Fixture 수정
 - 테스트 메소드명 알기 쉽게 변경 및 항상 새로 시작하도록 @After추가

jojoldu commented on src/main/java/nextstep/domain/RequestQuestionDto.java:
```diff
@@ -14,6 +14,9 @@
 
     public Question toEntity() {
-        return Question.builder().title(title).contents(contents).build();
+        return Question.builder()
```
피드백 반영 💯

jojoldu commented on src/main/java/nextstep/web/QnaController.java:
```diff
@@ -18,9 +17,8 @@
 @Controller
 @RequestMapping("/questions")
 @AllArgsConstructor
```
All이 아닌 required로 변환해주세요 :)

jojoldu commented on src/test/java/nextstep/web/api/ApiQnaAcceptanceTest.java:
```diff
@@ -49,55 +68,84 @@
+        String location = createResource(questionLocation + "/answers", "답변이애오 히히");
 
         Answer result = basicAuthTemplate().getForObject(location, Answer.class);
         softly.assertThat(result).isNotNull();
```
여기서 검증을 **controller 결과가 아닌 Repository에 저장된 정보**로 본다면 더욱 정확한 테스트가 될것 같습니다.
결국 등록/수정/삭제의 기능이 정상적으로 **DB에 반영**되는게 기능이기 때문입니다.
물론 컨트롤러에서 정상적인 값을 줬는지에 대한 테스트도 해야하는게 맞습니다 :)

저같은경우에 **get 기능은 컨트롤러 결과에 대한 검증**을, **post/put 등은 DB에 반영된건지에 대한 검증**을 합니다 :)

jojoldu reviewed (approved):
@isisrin 혜린님 피드백 반영 잘봤습니다!
빠른시간내에 해결해주셔서 감사합니다 :)

현재 남긴 코멘트는 **하면 좋고 안해도 괜찮은**정도의 코멘트이니 참고만 해주세요 :)

앞으로 스프링부트와 JPA에 관련된 코드를 작성하실때는 지금 남긴 코멘트도 꼭 참고해주세요 :)
고생하셨습니다!
